tiledb/sm/query/ast/query_ast.cc

```diff
@@ -0,0 +1,296 @@
+      attribute->type() != Datatype::STRING_ASCII &&
+      attribute->type() != Datatype::CHAR && (!attribute->var_size())) {
+    return Status_QueryConditionError(
+        "Value node attribute must have one value per cell for non-string "
+        "fixed size attributes: " +
+        field_name_);
+  }
+
+  // Ensure that the condition value size matches the attribute's
+  // value size.
+  if (attribute->cell_size() != constants::var_size &&
+      attribute->cell_size() != condition_value_size &&
+      !(attribute->nullable() && condition_value_view_.content() == nullptr) &&
+      attribute->type() != Datatype::STRING_ASCII &&
+      attribute->type() != Datatype::CHAR && (!attribute->var_size())) {
+    return Status_QueryConditionError(
+        "Value node condition value size mismatch: " +
+        std::to_string(attribute->cell_size()) +
+        " != " + std::to_string(condition_value_size));
+  }
+
+  // Ensure that the attribute type is valid.
+  switch (attribute->type()) {
+    case Datatype::ANY:
+      return Status_QueryConditionError(
+          "Value node attribute type may not be of type 'ANY': " + field_name_);
+    case Datatype::STRING_UTF8:
+    case Datatype::STRING_UTF16:
+    case Datatype::STRING_UTF32:
+    case Datatype::STRING_UCS2:
+    case Datatype::STRING_UCS4:
+      return Status_QueryConditionError(
+          "Value node attribute type may not be a UTF/UCS string: " +
+          field_name_);
+    default:
+      break;
+  }
+  return Status::Ok();
+}
+
+tdb_unique_ptr<ASTNode> ASTNodeVal::combine(
+    const tdb_unique_ptr<ASTNode>& rhs,
+    const QueryConditionCombinationOp& combination_op) {
+  std::vector<tdb_unique_ptr<ASTNode>> ast_nodes;
+  if (!rhs->is_expr()) {
+    ast_nodes.push_back(this->clone());
+    ast_nodes.push_back(rhs->clone());
+  } else {  // rhs is an expression node.
+    // lhs is a simple tree, rhs is a compound tree.
+    ast_nodes.push_back(this->clone());
+    if (rhs->get_node_combination_op() == combination_op) {
+      for (const auto& elem : rhs->get_node_children()) {
+        ast_nodes.push_back(elem->clone());
+      }
+    } else {
+      ast_nodes.push_back(rhs->clone());
+    }
+  }
+  return tdb_unique_ptr<ASTNode>(
+      tdb_new(ASTNodeExpr, std::move(ast_nodes), combination_op));
+}
+
+/** Value node getter methods */
+const std::string& ASTNodeVal::get_node_field_name() const {
+  return field_name_;
+}
+const ByteVecValue& ASTNodeVal::get_node_condition_value_data() const {
+  return condition_value_data_;
+}
+const UntypedDatumView& ASTNodeVal::get_node_condition_value_view() const {
+  return condition_value_view_;
+}
+const QueryConditionOp& ASTNodeVal::get_node_op() const {
+  return op_;
+}
+
+/** Expression node getter methods */
+const std::vector<tdb_unique_ptr<ASTNode>>& ASTNodeVal::get_node_children()
+    const {
+  throw std::runtime_error(
+      "ASTNodeVal::get_node_children: called get_node_children() on a value "
```

Comment:
I think something like "Cannot get children from a value node" would be enough. Similarly for the rest of getters.